### PR TITLE
set UchannelLO scale property as complex number

### PR DESCRIPTION
**Pull request by vvilpas**
### Summary

UchannelLO scale property was set as list instead of complex number when
instantiating from a fake backend. Now scale is converted to a complex number
if it is passed as list of two numbers to the constructor.

This causing issues like this one https://github.com/Qiskit/qiskit-aer/issues/1047

### CODE REVIEW

**Comment by taalexander** [comment]
In the provider, this is handled in the [JSONDecoder(Encoder)](https://github.com/Qiskit/qiskit-ibmq-provider/blob/c4738b52fd645b65dec38c7661ece2a7477c44a2/qiskit/providers/ibmq/utils/json_decoder.py#L68).

**Comment by vvilpas** [comment]
> In the provider, this is handled in the [JSONDecoder(Encoder)](https://github.com/Qiskit/qiskit-ibmq-provider/blob/c4738b52fd645b65dec38c7661ece2a7477c44a2/qiskit/providers/ibmq/utils/json_decoder.py#L68).

The json conf is loaded and used here (repeated in every backend that loads a `f_conf`): 

https://github.com/Qiskit/qiskit-terra/blob/d26f273d17ef2ce2a18be5d5eb51b86a684f6921/qiskit/test/mock/backends/valencia/fake_valencia.py#L38-L41

We could use the decoder you mention there or right in `from_dict` `PulseBackendConfiguration` class method, but there isn't one in terra, right? It won't catch direct instantiation of `UchannelLO` using a list instead of a complex number as `scale` (not very common usage I guess).

**Comment by taalexander** [comment]
Perhaps we could borrow this decoder just for the mock backends (duplicate in this package). The idea moving forward is the all providers should be responsible for their own serialization/decoding. Given this, the mock backends should likey be responsible for this.

**Comment by mtreinish** [comment]
> Perhaps we could borrow this decoder just for the mock backends (duplicate in this package). The idea moving forward is the all providers should be responsible for their own serialization/decoding. Given this, the mock backends should likey be responsible for this.

Yeah, I agree with this. This is something we should be handling directly in the `FakeBackend` class. The interface for `UchannelLO` takes in a complex, it's the provider/backend class's job to go between the wire format and what the terra interfaces expect. We'll also have a similar problem in other places where the iqx json payload differs from python types, complex is just the most obvious place this happens.

**Comment by vvilpas** [comment]
Makes sense. I'll take that approach.

**Comment by vvilpas** [comment]
@taalexander @mtreinish Ok, I made a few changes. Most of the mock fake backend classes (`FakeAlmaden`, `FakeArmong`, ...) share the same code without adding anything new, besides changing the location of the files from where `configuration`, `properties` and `defaults` (for open_pulse ones) are loaded. Thus, I've created a couple of abstract base classes where I've put all this common code: `FakeQasmBackend` and `FakePulseBackend` (which derives from the previous one), and almost all mock fake backends derive from them, changing only the location of json files.

The decode of the json configuration file happens in `FakeQasmBackend`, as `FakeBackend` expects already a `QasmBackendConfiguration`. We may move the decoding to the `FakeBackend` class, for example checking if `config` is a `dict` and do the decoding in that case and convert to a `QasmBackendConfiguration` or `PulseBackendConfiguration`, like:

```
if isinstance(configuration, dict):
    decode_backend_configuration(configuration)
    if configuration.get('open_pulse', False):
        configuration = PulseBackendConfiguration.from_dict(configuration)
    else:
        configuration = QasmBackendConfiguration.from_dict(configuration)
```

Let me know what you think.

**Review by taalexander — COMMENTED** [review]
I think these changes look good. Could you please verify that the automatic fetching and caching of backend data still works with the new class layout? See this script https://github.com/Qiskit/qiskit-terra/blob/master/tools/update_fake_backends.py. @mtreinish should also have a look given these significant changes.

**Comment by vvilpas** [comment]
> Could you please verify that the automatic fetching and caching of backend data still works with the new class layout? See this script https://github.com/Qiskit/qiskit-terra/blob/master/tools/update_fake_backends.py. 

Yeah, I've run the script and rerun all tests without any errors. Everything seems fine.

File: qiskit/test/mock/utils/json_decoder.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# that they have been altered from the originals.
+
+"""
+Utils to decode fake backend configurations from json
+"""
+
+from typing import Dict, Union, List
+
+
+def decode_backend_configuration(config: Dict) -> None:
```
Comment (taalexander) [inline qiskit/test/mock/utils/json_decoder.py:20]: It seems the accompanying code for the defaults/properties are missing (https://github.com/Qiskit/qiskit-ibmq-provider/blob/c4738b52fd645b65dec38c7661ece2a7477c44a2/qiskit/providers/ibmq/utils/json_decoder.py#L36-L65)?

Comment (vvilpas) [reply qiskit/test/mock/utils/json_decoder.py:54]: > It seems the accompanying code for the defaults/properties are missing (https://github.com/Qiskit/qiskit-ibmq-provider/blob/c4738b52fd645b65dec38c7661ece2a7477c44a2/qiskit/providers/ibmq/utils/json_decoder.py#L36-L65)?

I've added them and used at the appropriate places.

**Review by taalexander — APPROVED** [review]
This looks great! Thanks for all the hard work 👍 . This initial small change turned into a meaningful improvement to our test infrastructure.

**Pull request merged**